core/src/main/java/com/microsoft/applicationinsights/internal/statsbeat/CustomDimensions.java

```diff
@@ -0,0 +1,114 @@
+
+    private static CustomDimensions instance;
+    private String resourceProvider;
+    private String operatingSystem;
+    private String customerIkey;
+    private String version;
+    private String runtimeVersion;
+    private ConcurrentMap<String, String> properties;
+    private static final Object lock = new Object();
+    
+    public static CustomDimensions getInstance() {
+        synchronized (lock) {
+            if (instance == null) {
+                instance = new CustomDimensions();
+            }
+        }
+        return instance;
+    }
+    
+    private CustomDimensions() {
+        initialize();
+    }
+
+    private void initialize() {
+        String sdkVersion = PropertyHelper.getQualifiedSdkVersionString();
+        if (sdkVersion.startsWith("awr")) {
+            resourceProvider = RP_APPSVC;
+            operatingSystem = OS_WINDOWS;
+        } else if (sdkVersion.startsWith("alr")) {
+            resourceProvider = RP_APPSVC;
+            operatingSystem = OS_LINUX;
+        } else if (sdkVersion.startsWith("kwr")) {
+            resourceProvider = RP_AKS;
+            operatingSystem = OS_WINDOWS;
+        } else if (sdkVersion.startsWith("klr")) {
+            resourceProvider = RP_AKS;
+            operatingSystem = OS_LINUX;
+        } else if (sdkVersion.startsWith("fwr")) {
+            resourceProvider = RP_FUNCTIONS;
+            operatingSystem = OS_WINDOWS;
+        } else if (sdkVersion.startsWith("flr")) {
+            resourceProvider = RP_FUNCTIONS;
+            operatingSystem = OS_LINUX;
+        } else if (sdkVersion.startsWith(LANGUAGE)) {
+            resourceProvider = UNKNOWN;
+        }
+
+        if (operatingSystem == null) {
+            if (SystemInformation.INSTANCE.isWindows()) {
+                operatingSystem = OS_WINDOWS;
+            } else if (SystemInformation.INSTANCE.isUnix()) {
+                operatingSystem = OS_LINUX;
+            } else {
+                operatingSystem = OS_UNKNOW;
+            }
+        }
+
+        customerIkey = TelemetryConfiguration.getActive().getInstrumentationKey();
+        version = sdkVersion.substring(sdkVersion.lastIndexOf(':') + 1);
+        runtimeVersion = System.getProperty("java.version");
+        
+        properties = new ConcurrentHashMap<>();
+        properties.put(CUSTOM_DIMENSIONS_RP, resourceProvider);
+        properties.put(CUSTOM_DIMENSIONS_ATTACH_TYPE, ATTACH_TYPE_CODELESS);
+        if (customerIkey != null) { // Unit test
+            properties.put(CUSTOM_DIMENSIONS_CIKEY, customerIkey);
+        }
+        properties.put(CUSTOM_DIMENSIONS_RUNTIME_VERSION, runtimeVersion);
+        properties.put(CUSTOM_DIMENSIONS_OS, operatingSystem);
+        properties.put(CUSTOM_DIMENSIONS_LANGUAGE, LANGUAGE);
+        properties.put(CUSTOM_DIMENSIONS_VERSION, version);
+    }
+
+    public ConcurrentMap<String, String> getProperties() {
+        return properties;
+    }
+
+    public static synchronized void reset() {
```

Comment:
also, is synchronized needed?
```suggestion
    // only used by tests
    static synchronized void reset() {
```